### ISSUE TITLE
SPFX webpart serving old files versions in Chrome

### ISSUE THREAD

**Issue by Ryan365Apps** [SharePoint/sp-dev-docs#7664]
### Target SharePoint environment

SharePoint Online

### What SharePoint development model, framework, SDK or API is this about?

💥 SharePoint Framework

### Developer environment

Windows

### What browser(s) / client(s) have you tested

- [ ] 💥 Internet Explorer
- [X] 💥 Microsoft Edge
- [X] 💥 Google Chrome
- [ ] 💥 FireFox
- [ ] 💥 Safari
- [ ] mobile (iOS/iPadOS)
- [ ] mobile (Android)
- [ ] not applicable
- [ ] other (enter in the "Additional environment details" area below)

### Additional environment details

- browser version
Google Chrome - Version 97.0.4692.99 (Official Build) (64-bit)
Microsoft Edge -  Version 97.0.1072.62 (Official build) (64-bit)
- SPFx version
1.12.1
- Node.js version
10.13

### Describe the bug / error

I'm facing a strange error where SPFX web part is serving an old version of the CDN files when using a standard refresh in chrome but works fine when doing a hard reload. 

The problem appears again if another standard refresh is shown.

Gif showing the same page in both browsers side by side.

![RefreshExample](https://user-images.githubusercontent.com/50838643/150508257-476424e1-f45b-4903-9b0f-8060790b61f2.gif)

**Important to note that this has been reported to have occurred across a number of my customer's Production tenants and I can provide the details of those privately.

I believe this is similar to this issue 
[SPFx loading old version files and sometimes from different environment #7008](https://github.com/SharePoint/sp-dev-docs/issues/7008)

Apologies if more information is needed on build/environment etc but my assumption is that this isn't a specific config of my setup. 

Thanks

### Steps to reproduce

1. Add custom web part in Chrome & Edge

### Expected behavior

Same consistent experience in Chrome and Edge

**Comment by Ekaterina-Balashchenko-ITechArt** [comment]
The same situation in one tenant 2 days ago and today in another tenant ((

**Comment by uladzislau-loputs-itechart** [comment]
> The same situation in one tenant 2 days ago and today in another tenant ((

Adding details: one tenant is using O365 CDN, another one just uses app catalog for assets

**Comment by AJIXuMuK** [comment]
From #7662:

>  have checked that recently on some of the tenants my spfx packages are returned to their initial versions, due to this their functionality also showing which I implemented and deployed in first version and everything which I deployed in my latest versions are vanished.
It reflects on some of the tenants for me.
My current version is 1.0.0.2 but on app catalog is changed to 1.0.0.1.
It happened with my all .sppkg packages on multiple tenants also.
Right now I resolved it by re-deploying my latest packages on the app catalog for all tenants.

**Comment by Ryan365Apps** [comment]
Had tried that with no success. My example above shows exactly the same site page open in two browsers, so the deployed package is the same. 

It seems that the older files are persisting in the cache storage, and the standard refresh is still serving those cached files  - I compared this with files served via "empty cache and hard reload" and all the metadata was identical, but the content was different. 

I resolved this for one of my dev tenants by _manually_ finding the entry in the cache of chrome and deleting the file.

I can't do this for over 40 customer tenants so I'm hoping for a better resolution

![image](https://user-images.githubusercontent.com/50838643/150544039-390b9d1d-b105-4f70-8780-276f0e06394b.png)

**Comment by tandy-1000-hx** [comment]
Our company's tenant experienced a similar issue this morning. All custom SPFX developed web parts reverted back to the first version that was uploaded to our tenant's SharePoint app catalog.  The old version of the webpart showed in Edge, Chrome and Firefox. Emptying the cache did not resolve it.  The SharePoint app catalog correctly shows the latest version number (i.e. 1.9, 1.12, etc) of each web part, but the actual version that loads on SharePoint is v1.0.

For the webparts deployed tenant-wide, I was able to go to the SharePoint app catalog, click one of the web parts, then click "Deploy" at the top and the latest version appeared immediately in SharePoint.  Which begs the question: why did they revert to the initial version to begin with?

**Comment by patmill** [comment]
This is all super weird.  For people that are facing this, can you go to the manifests list (sites/appcatalog/Lists/ComponentManifests/) and find the components that are misbehaving?  If you look at the ComponentManifest, there should be a path property that has the name of the js file.  Is that correct (pointing at the new version) or incorrect (pointing at the old one)?  Additionally, if you go to the appropriate folder in sites/appcatalog/ClientSideAssets and drill into your solution, are the new or old files there?

Finally, can you look at the version history of these and see if something edited / updated them?

**Comment by patmill** [comment]
Also, how are your js files normally getting deployed?
These URLs don't seem standard (they aren't in clientsideassets)
![image](https://user-images.githubusercontent.com/3767881/150599228-60729f6b-05e7-4a87-9465-1a52a4d14aa2.png)

**Comment by patmill** [comment]
One more question - is there any chance that you modified the componentmanifests list directly, and changed the paths?

Something seems to have triggered deploying the package (maybe an older version?) which takes the entries in the sppkg file and updates the componentmanifests and clientsideassets folder.  If you have direct links to files outside the clientsideassets folder, then the manifest might be pointing to things that are no longer there.

**Comment by tandy-1000-hx** [comment]
@patmill , thank you for following up and for the acknowledgement and explanation of the issue! I was going in circles trying to figure out how it happened.  I see Microsoft has now sent out a bulletin for it in the tenant's Service Health admin center.
Thanks again for following up!

**Comment by mbice** [comment]
Hi,

We're experiencing this same issue on a bunch of tenants, but our issue seems to be targeted more on library components and re-deploying doesn't resolve it.

We load our library components using SPComponentLoader.loadComponentById() and this was working fine until we noticed this bug yesterday. Even with fresh deployments of the library component sppkg files it persists (I even tried deleting them entirely and re-deploying).

On a cold load, no cache, they are loaded just fine. Subsequent page loads from cache and the SP component loader can't find the manfiests (just comes back null).

Could there be a separate issue with library components missing from cache when using the SPComponentLoader? Older tenants where we had previously deployed these packages are loading ok. It seems to be tenants we've spun up in the last week or so that are experiencing this issue.

Thanks!

**Comment by Ryan365Apps** [comment]
We are still experiencing this issue - one of my developer tenants and a customer's production tenant (tenant details available if needed). 

As @mbice states "cold load, no cache" works fine but any other page load uses much older cached content. Checked in and re-deployed solution does not work.

The admin center message SP316684 is stated as "service restored"

**Comment by mbice** [comment]
We're seeing this issue spread to more tenants and not just new ones we've created in the past few weeks. Not sure if re-deploying packages is triggering the library component cache load issue or if it's happening automatically.

@patmill is the library component part of this issue known on your side yet?

**Comment by patmill** [comment]
To help diagnose, can you try hitting your page with ?sw=bypass   and see if the problem goes away?

**Comment by mbice** [comment]
@patmill adding ?sw=bypass on page load does make the problem go away, but then it persists again on subsequent page loads once the querystring parameter is removed.

Thanks!

**Comment by sherpas-par** [comment]
> o help diagnose, can you try hitting your page with ?sw=bypass and see if the problem goes away?

Tried your suggestion and the problem is resolved when adding ?sw=bypass but when reloading the page the problem is back since the querystring is removed.

**Comment by till-llit** [comment]
can reproduce this issue in both, standard and targeted release.

CTRL+F5 or ?sw=bypass is "healing" the issue for one load cycle.

**Comment by arunakshi** [comment]
This effected also our production tenant since yeaterday mid-night.

@patmill 
We checked our versions in ComponentManifests and also Clientsideassets. They look fine. I also did re-deployment in our dev Environment where we have similar issue, it didnt help us. Any hints please how we dig further.

**Comment by AJIXuMuK** [comment]
Just to update the thread.
The team has made fixes that should unblock these scenarios. They will be rolled out WW in the next 2 weeks.
Additionally what you could to to unblock right now:
- Return `Access-Control-Expose-Headers: Date` to allow service worked read the headers for the pre-flight
- set `cache-control` to `no-cache,max-age=0`

This should unblock you.

**Comment by desaiprerak** [comment]
Could you elaborate on where we have to add this work around ?

**Comment by AJIXuMuK** [comment]
These should be response headers for you custom CDN location.

**Comment by mahmoudhhassan** [comment]
@AJIXuMuK if you are using default SPFx setup this will be the MS Public CDN that you don't have control over!

**Comment by AJIXuMuK** [comment]
@mahmoudhhassan - if you look at the sample that @Ryan365Apps posted above - this is not MS CDN:
<img width="460" alt="image" src="https://user-images.githubusercontent.com/17036219/151584015-78e54489-b303-4cc1-80a5-418aa2821fc8.png">

**Comment by mbice** [comment]
@AJIXuMuK if it helps with troubleshooting I've been able to reproduce this issue with or without the M365 CDN enabled.

**Comment by AJIXuMuK** [comment]
@mbice - just to confirm: you don't use custom CDN and custom deployment configuration for SPFx solution, correct?
You're just deploying all the assets to the app catalog/client side assets, and serving it from there. Correct?

**Comment by arunakshi** [comment]
@AJIXuMuK our case is exactly what you mentioned. We do not use custom CDN and deploying our assets to app catalog/clientsideassets. 

We cannot use cache-control since we use default SPFX setup.

**Comment by mbice** [comment]
@AJIXuMuK we typically use M365 CDN, but as part of my testing for this issue I disabled it on one of our tenants and the issue was still there. No matter how CDN is configured (on or off) on cached page loads I see the following console error when library component manifests are attempted to be loaded via SPComponentLoader.loadComponentById().

![image](https://user-images.githubusercontent.com/985283/151590135-f4c7e3c8-ebca-4687-836a-1bab48eb6105.png)

Looks like it's failing when retrieving the library component manifest. Manifest comes back null and then the component fails to load. On non-cached load this method call works fine.

![image](https://user-images.githubusercontent.com/985283/151590574-6c2c5730-236e-4a6d-9a99-97df5181e225.png)

Hope that helps!

**Comment by AJIXuMuK** [comment]
@arunakshi @mbice - does the `sw=bypass` solve the issue for you too?

**Comment by mbice** [comment]
@AJIXuMuK yes adding that to the querystring fixes the issue, but the issue returns on subsequent page loads once that querystring parameter is removed.

Thanks

**Comment by arunakshi** [comment]
@AJIXuMuK, yes it is also working for me if we add sw=bypass and also querystring parameter on subsequent page loads like @mbice.

**Comment by AJIXuMuK** [comment]
Thanks for the confirmation.
So, this GH issue currently consists of 3:
- one related to updater and checked out solutions. **Has been resolved**
- Issue with incorrectly cached/loaded assets from custom CDNs. **The fix will roll out in the next 2 weeks, use the workaround I posted above**
- Issue with incorrectly loaded manifest. **This has been fixed and should be rolled out next week. No workaround for this one**
So, currently my only suggestion is to wait until the fixes will be rolled out and see if there are any unfixed scenarios left.

**Comment by till-llit** [comment]
@AJIXuMuK : you might want to allow #7685 to be investigated seperatly, The issue with SPComponentLoader might be an isolated issue since 28.1 and not releated to this issue here. Even thought it can be resolved temporarly with CTRL+F5 or ?sw=bypass the root cause is not an incorrectly loaded manifest.

**Comment by AJIXuMuK** [comment]
@till-llit - there are multiple issues with the Service Worker that are tracked here. And there are multiple fixes that have been made.
If the issue is not resolved when the fix is rolled out - you can definitely open a separate one. But I believe it will be fixed with the changes to the service worker.

**Comment by wobba** [comment]
@AJIXuMuK please update on when fixes will roll out as we get more and more issues related to dynamic loading of SPFx libs for the PnP Modern Search web parts. Seems a recent SPFx change caused this error with 3p solutions dynamically loading libs using `SPComponentLoader`.

**Comment by AJIXuMuK** [comment]
@wobba if everything goes ok one of the fixes should be rolled out by Thursday this week, and another one - by end of next week

**Comment by till-llit** [comment]
@wobba : the issue with SPComponentLoader was fixed for both, our Standard Release and Targeted Release tenant, tonight.

**Comment by FredrikEkstroem** [comment]
@wobba my issues with SPComponentLoader in PNP search webparts are not reproducable today, so looks like last nights fix worked

**Comment by Ryan365Apps** [comment]
> * Issue with incorrectly cached/loaded assets from custom CDNs. **The fix will roll out in the next 2 weeks, use the workaround I posted above**

Any update on the progress of rolling out this fix? 

Please can you advise on how to apply work-around for CDN response headers? 
This is how I have configured it but it has no impact

![image](https://user-images.githubusercontent.com/50838643/153005019-c782ecb7-bac1-4dff-902c-72266519cbd2.png)

**Comment by AJIXuMuK** [comment]
Hi @Ryan365Apps - it's still rolling out.

For the headers - they look correct. If I request the file from your screenshot (https://365apps.azureedge.net/cwp/teams-web-part_a4d6041218591ad62db3.js) - I see the headers in the Response.

**Comment by Ryan365Apps** [comment]
Hi, We still have this issue with some of our customer tenants. Can we get an ETA on completion, please?

**Comment by AJIXuMuK** [comment]
The fixes are rolled out on 02/14.

If you're still experiencing issues on _some_ tenants - please, check if there is anything special on that tenants from your app perspective too.

**Comment by arunakshi** [comment]
We are facing again the following error in our DEV tenant which was not this case earlier. In the background, we try to load a library component using  SPComponentLoader.loadComponentById(). 
![image](https://user-images.githubusercontent.com/12812202/155087008-1d1cf70d-8294-4797-a66c-4b2be845ca53.png)

This is strange. It was working fine for us last couple of days.

**Comment by arunakshi** [comment]
@AJIXuMuK - No this is the later one. Okay, I will open a separate issue for that. Thanks!

**Comment by AJIXuMuK** [comment]
@Ryan365Apps - do you still experience the issue?
If yes - could you provide the details?

**Comment by AJIXuMuK** [comment]
Hi @Ryan365Apps - thank you for checking that.

Please, send the tenants info as well as URLs you're trying to reach to aterentiev@microsoft.com

**Comment by rmillener** [comment]
This is currently happening in at least two of our customer tenants in the Australian region with the same result. As a workaround we have re-packaged our solutions with new IDs. 

In one tenancy it is loading a debug bundle that was replaced with a production version in February.
